### PR TITLE
VP-1480 Create IIndexDocumentRegistrar

### PR DESCRIPTION
**Pull request by AliveMen**
Create IIndexDocumentRegistrar
Use IndexDocumentRegistrar instead IEnumerable<IndexDocumentConfiguration>

### CODE REVIEW

File: src/VirtoCommerce.SearchModule.Data/Services/IndexingManager.cs [hunk 1]
```diff
@@ -18,12 +18,12 @@ namespace VirtoCommerce.SearchModule.Data.Services
...
     {
         private readonly ISearchProvider _searchProvider;
-        private readonly IEnumerable<IndexDocumentConfiguration> _configs;
+        private readonly IIndexDocumentRegistrar _configs;
         private readonly ISettingsManager _settingsManager;
         private readonly IIndexingWorker _backgroundWorker;
         private readonly SearchOptions _searchOptions;
 
-        public IndexingManager(ISearchProvider searchProvider, IEnumerable<IndexDocumentConfiguration> configs, IOptions<SearchOptions> searchOptions,
+        public IndexingManager(ISearchProvider searchProvider, IIndexDocumentRegistrar configs, IOptions<SearchOptions> searchOptions,
```
Comment (yecli) [inline src/VirtoCommerce.SearchModule.Data/Services/IndexingManager.cs]: Need to rename `configs`

File: src/VirtoCommerce.SearchModule.Core/Model/IndexDocumentConfiguration.cs [hunk 2]
```diff
@@ -1,11 +1,56 @@
+using System;
 using System.Collections.Generic;
+using VirtoCommerce.Platform.Core.Common;
 
 namespace VirtoCommerce.SearchModule.Core.Model
 {
+    public class IndexDocumentConfigurationBuilder
```
Comment (yecli) [inline src/VirtoCommerce.SearchModule.Core/Model/IndexDocumentConfiguration.cs]: Need to move Builder and Extensions to separate files

File: src/VirtoCommerce.SearchModule.Core/Services/IIndexDocumentRegistrar.cs [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+        IEnumerable<IndexDocumentConfiguration> GetIndexDocumentConfigurations();
+
+        /// <summary>
+        /// Gets list of registered IndexDocumentConfiguration filtered by document type (e.g. KnownDocumentTypes.Product).
+        /// </summary>
+        /// <param></param>
+        /// <param name="documentType"></param>
+        /// <returns>IEnumerable<IndexDocumentConfiguration></returns>
+        /// <exception cref="InvalidOperationException">Thrown when IndexDocumentConfigurations with the requested document type is not registered. </exception>
+        IEnumerable<IndexDocumentConfiguration> GetIndexDocumentConfigurations(string documentType);
```
[the remark this answers is not in the document]
Comment (AliveMen) [reply src/VirtoCommerce.SearchModule.Core/Services/IIndexDocumentRegistrar.cs]: Changed to GetIndexDocumentConfiguration

File: src/VirtoCommerce.SearchModule.Data/Services/IndexDocumentRegistrar.cs [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+using System.Linq;
+using VirtoCommerce.Platform.Core.Common;
+using VirtoCommerce.SearchModule.Core.Model;
+using VirtoCommerce.SearchModule.Core.Services;
+
+namespace VirtoCommerce.SearchModule.Data.Services
+{
+    public class IndexDocumentRegistrar : IIndexDocumentRegistrar
+    {
+        private readonly List<IndexDocumentConfiguration> _indexDocument = new List<IndexDocumentConfiguration>();
```
[the remark this answers is not in the document]
Comment (AliveMen) [reply src/VirtoCommerce.SearchModule.Data/Services/IndexDocumentRegistrar.cs]: done

File: src/VirtoCommerce.SearchModule.Data/Services/IndexDocumentRegistrar.cs [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+        public IEnumerable<IndexDocumentConfiguration> GetIndexDocumentConfigurations()
+        {
+            return _indexDocument;
+        }
+
+        public IEnumerable<IndexDocumentConfiguration> GetIndexDocumentConfigurations(string documentType)
+        {
+            var indexDocumentConfigurations = _indexDocument.Where(x => string.Equals(x.DocumentType, documentType, StringComparison.InvariantCultureIgnoreCase)).ToArray();
```
[the remark this answers is not in the document]
Comment (AliveMen) [reply src/VirtoCommerce.SearchModule.Data/Services/IndexDocumentRegistrar.cs]: done

File: src/VirtoCommerce.SearchModule.Data/Services/IndexDocumentRegistrar.cs [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+            if (indexDocumentConfigurations.IsNullOrEmpty())
+            {
+                throw new InvalidOperationException($"IndexDocumentConfiguration with document type :'{documentType}' not exists ");
```
[the remark this answers is not in the document]
Comment (AliveMen) [reply src/VirtoCommerce.SearchModule.Data/Services/IndexDocumentRegistrar.cs]: fixed

File: src/VirtoCommerce.SearchModule.Data/Services/IndexDocumentRegistrar.cs [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+            {
+                throw new InvalidOperationException($"IndexDocumentConfiguration with document type :'{documentType}' already exists ");
+            }
+
+            _indexDocument.Add(new IndexDocumentConfiguration {DocumentType = documentType, DocumentSource = documentSource});
+        }
+
+        public void RegisterRelatedSource(string documentType, IndexDocumentSource documentSource)
+        {
+            foreach (var configuration in GetIndexDocumentConfigurations(documentType))
```
[the remark this answers is not in the document]
Comment (AliveMen) [reply src/VirtoCommerce.SearchModule.Data/Services/IndexDocumentRegistrar.cs]: fixed

File: src/VirtoCommerce.SearchModule.Data/Services/IndexingManager.cs [hunk 8]
```diff
@@ -90,7 +91,7 @@ public virtual async Task IndexAsync(IndexingOptions options, Action<IndexingPro
                 await _searchProvider.DeleteIndexAsync(documentType);
             }
 
-            var configs = _configs.Where(c => c.DocumentType.EqualsInvariant(documentType)).ToArray();
+            var configs = _configs.GetIndexDocumentConfigurations(documentType);
```
[the remark this answers is not in the document]
Comment (AliveMen) [reply src/VirtoCommerce.SearchModule.Data/Services/IndexingManager.cs]: Duplication removed

File: src/VirtoCommerce.SearchModule.Data/Services/IndexingManager.cs [hunk 9]
```diff
@@ -101,7 +102,7 @@ public virtual async Task IndexAsync(IndexingOptions options, Action<IndexingPro
         public virtual async Task<IndexingResult> IndexDocumentsAsync(string documentType, string[] documentIds)
         {
             // Todo: reuse general index api?
-            var configs = _configs.Where(c => c.DocumentType.EqualsInvariant(documentType)).ToArray();
+            var configs = _configs.GetIndexDocumentConfigurations(documentType);
```
[the remark this answers is not in the document]
Comment (AliveMen) [reply src/VirtoCommerce.SearchModule.Data/Services/IndexingManager.cs]: Duplication removed

**Pull request closed without merging**